core/ledger/kvledger/upgrade_dbs.go

```diff
@@ -24,29 +33,44 @@ func UpgradeDBs(rootFSPath string) error {
 	defer fileLock.Unlock()
 
 	logger.Infof("Ledger data folder from config = [%s]", rootFSPath)
+	dbPath := LedgerProviderPath(rootFSPath)
+	db := leveldbhelper.CreateDB(&leveldbhelper.Conf{DBPath: dbPath})
+	db.Open()
+	defer db.Close()
 
-	// upgrades idStore (ledgerProvider) and drop databases
-	if err := UpgradeIDStoreFormat(rootFSPath); err != nil {
-		return err
+	if emptydb, _ := db.IsEmpty(); emptydb {
+		return errors.Errorf("ledger database %s is empty. Make sure core.yaml has the correct ledger path.", dbPath)
 	}
 
-	if err := dropDBs(rootFSPath); err != nil {
+	idStore := &idStore{db, dbPath}
+	format, err := idStore.getFormat()
+	if err != nil {
+		return err
+	}
+	if bytes.Equal(format, []byte(dataformat.Version20)) {
+		logger.Info("Ledger data format is current, do nothing.")
+		return nil
+	}
+	if format != nil {
+		err = &dataformat.ErrVersionMismatch{
+			ExpectedVersion: "",
+			Version:         string(format),
+			DBInfo:          fmt.Sprintf("leveldb for channel-IDs at [%s]", dbPath),
+		}
```

Comment:
> We had this check deliberately in the function idStore.upgradeFormat. So that we do not accidentally end-up using that function without a precondition check somewhere. I suggest to keep it there for sure. I am fine, if you want to check here again.

I suggested this to be removed from `idStore.upgradeFormat()` because it introduces code duplication, execution duplication, and side-effect in the function `upgradeFormat()`. Currently, `idStore.upgradeFormat()` is used only by the `UpgradeDBs()`. I don't see a strong justification for code duplication, execution duplication, and side-effect but I will step aside if you see it.

Reply:
I gave the justification in my comment - which is code robustness. A function better not be knowing that how it is being used and hence preconditions be checked inside. 
There are situations when we cannnot avoid that e.g., assume that the lock has been acquired by the caller. But there you still prefer to keep function name either xxxwithoutLock or specify in the comments.
Here, repeating an if check twice in the code flow is not such a big problem if its helping in early return as well as code strength.

What side-effect are you referring to?

Reply:
For the above-mentioned justification only, I mentioned it is not strong enough (for me). At the same time, if you are okay with it, I have no problem in having the following.  

- at least 10 lines are duplicated.
- in the UpgradeDBs() execution flow, these 10 lines occurs twice.
- only one functions call updateFormat. Hence, I don't know about the robustness. I also couldn't speculate what other future functions would use it and what are the probabilities that we would miss checking the current format before calling updateFormat. 
- `updateFormat()` is more like `updateFormatIfNotCurrent()`

Reply:
I moved the checking per review comment mainly because it is a package-only method - we own the caller code to do the checking before calling. However, I agree with Manish to also keep the checking in idStore.upgradeFormat. 
I propose to add a func idStore.checkFormatUpgradeEligibility that will be called by UpgradeDBs and idStore.ugradeFormat. This way idStore.getFormat is not needed any more.

Reply:
As the majority feels the other way, I will step aside. Sorry, @wenjianqiao for the extra work incase you have to revert the changes.